README.md
# Vaultwarden backup

![GitHub](https://img.shields.io/github/license/malinkinsa/vaultwarden-backup)

## About

This is an application for creating backup Vaultwarden files and databases.

## Usage

### Configuration

The application requires a configuration file in TOML format to specify the Vaultwarden data directory, database connection parameters and the backup location. The default location is ```config.toml``` in the same directory, but this can be configured by setting the CONFIG_PATH env variable to whatever path you would like. An example configuration file ```config.toml``` is provided in the project directory.

Configuration values are as follows:

| Name             | Type     | Optional | Default value                                                                  | Description                                                                                                                                                |
|------------------|----------|----------|--------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------|
| vaultwarden_data | String   |          |                                                                                | The path to the Vaultwarden data directory                                                                                                                 |
| backup_location  | String   |          |                                                                                | The path to the directory where the backup will be stored                                                                                                  |
| exclude_files    | String   | Optional |                                                                                | The list consists of files that should be excluded from the backup process by checking if they contain a certain substring, rather than by their full name |
| encrypt_data     | Boolean  | Optional | false                                                                          | The key responsible for determining whether to protect the archive with a password. You need to have the 7zip (7zz binary) package installed.              |
| encrypt_key      | String   | Optional |                                                                                | The password that will be set for the archive.                                                                                                             |   
| db               | Database |          |                                                                                | Block containing database configuration details                                                                                                            |
| db_type          | String   |          |                                                                                | The type of the database (postgresql, mysql, mariadb, or sqlite)                                                                                           |
| username         | String   | Optional |                                                                                | The username to connect to the database                                                                                                                    |
| password         | String   | Optional |                                                                                | The password to connect to the database                                                                                                                    |
| db_name          | String   | Optional |                                                                                | The name of the database                                                                                                                                   |
| host             | String   | Optional |                                                                                | The hostname/IP address of the database                                                                                                                    |
| port             | Integer  | Optional | Postgresql: 5432<br/> MySQL/MariaDB: 3306<br/> SQLite: vaultwarden_data folder | The port to connect to the database                                                                                                                        |

### Launch

- Download the binary from [Releases](https://github.com/malinkinsa/vaultwarden-backup/releases) or build it from source;
- Configure the ```config.toml``` file in the same directory or another directory and specify its location via the ```CONFIG_PATH``` environment variable;
- Launch the binary by running ```./vaultwarden-backup```


## Changelog

Check out the [CHANGELOG](CHANGELOG.md) file.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.

CHANGELOG.md
# 0.2.1

## Breaking changes
- None

## New features
- None

## Other changes
- There is no longer a need for 7zip. The creation of a secure archive is now done using the zip-rs library.

# 0.2.0

## Breaking changes
- On output, a single archive file is generated, containing the database dump and system files in the form of zip/tar.gz. 
- The archive filename has been modified.

## New features
- Added the ability to create a password-protected archive.

## Other changes
- None

# 0.1.1

## Breaking changes
- None

## New features
- The ability to specify which files to exclude from the backup process has been added. Additionally, all files related to SQLite are always excluded from backup, as the database itself is backed up separately in the first step.

## Other changes
- None

# 0.1.0

## Breaking changes
- None

## New features
- Added functionality for backup both databases (PostgreSQL, MySQL, SQLite) and files.

## Other changes
- None

src/file.rs
use std::error::Error;
use flate2::Compression;
use flate2::write::GzEncoder;
use tar::Builder;
use std::fs;
use std::fs::File;
use std::path::Path;
use walkdir::WalkDir;

use crate::middleware::{delete_temp_dir, zip_dir};

pub fn create_temp_dir(backup_location: &str, formatted_datetime: &str) -> Result<String, Box<dyn Error>> {
    let temp_dir = match backup_location.ends_with('/') {
        true => format!("{}{}/", backup_location, formatted_datetime),
        false => format!("{}/{}/", backup_location, formatted_datetime),
    };

    fs::create_dir(&temp_dir)?;
    Ok(temp_dir)
}

pub fn create_archive_with_vw_files(backup_location: &String, folder_for_archive: &String, exclude_files: Vec<String>) -> Result<(), Box<dyn Error>> {
    let archive_file = File::create(format!("{}data.tar.gz", backup_location))?;
    let encoder = GzEncoder::new(archive_file, Compression::default());

    let mut archive_builder = Builder::new(encoder);

    let target_dir = Path::new(folder_for_archive);
    let target_entries = fs::read_dir(target_dir)?;

    let mut exclude_files = exclude_files;
    exclude_files.push("sqlite3".to_string());

    for entry in target_entries {
        let entry = entry?;
        let entry_path = entry.path();
        let relative_path = entry_path.strip_prefix(target_dir).unwrap();

        if entry_path.is_file() {
            if let Some(target_file) = entry.file_name().to_str() {
                if !exclude_files.iter().any(|f|target_file.contains(f.trim())) {
                    if let Err(error) = archive_builder.append_file(relative_path.to_str().unwrap(), &mut File::open(&entry_path)?) {
                        return Err(Box::new(error))
                    }
                } else {
                    eprintln!("The backup process has excluded the following files: {}", entry_path.display());
                }
            }
        } else if let Err(error) = archive_builder.append_dir_all(relative_path.to_str().unwrap(), &entry_path) {
            return Err(Box::new(error));
        }
    }

    Ok(())
}

pub fn create_archive(
    archive_type: bool,
    archive_key: String,
    directory_path: &str
) -> Result<(), Box<dyn Error>> {

    if archive_type {
        let protected_archive = create_password_protected_archive(directory_path, &archive_key);
        if let Err(error) = protected_archive {
            eprintln!("(!) {}", error);
        }
    } else {
        let unprotected_archive: Result<(), Box<dyn Error>> = create_unprotected_archive(directory_path);

        if let Err(error) = unprotected_archive {
            eprintln!("(!) {}", error);
        }
    }
    Ok(())
}

fn create_unprotected_archive(directory_path: &str) -> Result<(), Box<dyn Error>> {
    let archive_file = File::create(format!("{}.tar.gz", directory_path.trim_end_matches('/')))?;
    let encoder = GzEncoder::new(archive_file, Compression::default());

    let mut archive_builder = Builder::new(encoder);

    if let Err(error) = archive_builder.append_dir_all(".", directory_path) {
        return Err(Box::new(error))
    } else {
        eprintln!("A backup archive has been created.");
        if let Err(err) = delete_temp_dir(directory_path.trim_end_matches('/')) {
            eprintln!("Failed to delete temporary directory: {}", err)
        }
    }

    Ok(())
}

fn create_password_protected_archive(
    directory_path: &str,
    archive_key: &str
) -> zip::result::ZipResult<()> {

    let walkdir = WalkDir::new(directory_path);
    let archive_file = File::create(format!("{}.zip", directory_path.trim_end_matches('/'))).unwrap();

    let items = walkdir.into_iter();

    zip_dir(&mut items.filter_map(|e| e.ok()), directory_path, archive_file, archive_key)?;

    if let Err(err) = delete_temp_dir(directory_path.trim_end_matches('/')) {
        eprintln!("Failed to delete temporary directory: {}", err)
    }

    Ok(())

}

src/main.rs
use chrono::{DateTime, Local};
use std::process::exit;

use crate::config::Config;
use crate::file::create_archive;

mod config;
mod database;
mod file;
mod middleware;

fn main() {
    let config:Config = Config::new();
    if let Ok(backup_location) = config.get_backup_location() {
        let start_datetime: DateTime<Local> = Local::now();
        let formatted_datetime: String = start_datetime.format("%d-%m-%Y_%H-%M").to_string();
        let datetime: String = start_datetime.format("%Y-%m-%d-%H:%M").to_string();

        eprintln!("Backup process started at {}\nCreating a temporary directory", start_datetime.format("%Y-%m-%d %H:%M:%S"));
        let temp_dir = file::create_temp_dir(backup_location, &formatted_datetime).unwrap();

        let db_backup = database::database_backup(
            config.get_db_type(),
            config.db_connection_string(),
            &temp_dir,
            datetime
        );

        if let Err(error) = db_backup {
            eprintln!("(!)Error during database backup process: {}", error);
            exit(1)
        } else {
            eprintln!("Starting file backup")
        }

        let files_backup = file::create_archive_with_vw_files(
            &temp_dir,
            &config.vaultwarden_data,
            config.get_exclude_files()
        );

        if let Err(error) = files_backup {
            eprintln!("(!)Error during files backup process: {}", error);
            exit(1)
        } else {
            eprintln!("Files backup successfully completed")
        }

        let archive = create_archive(
            config.get_encrypt_status(),
            config.get_encrypt_key(),
            &temp_dir,
        );

        if let Err(error) = archive {
            eprintln!("(!) {}", error);
            exit(1)
        }

        let end_datetime: DateTime<Local> = Local::now();

        eprintln!("Backup process ended at {}", end_datetime.format("%Y-%m-%d %H:%M:%S"));
    } else {
        let error = config.get_backup_location().unwrap_err();
        eprintln!("(!)Error: {}", error)
    }
}


src/database.rs
use std::error::Error;
use std::fs::File;
use std::process::{Command, Output};

pub fn database_backup(db_type: String, connection_string: String, backup_location: &String, datetime: String) -> Result<(), Box<dyn Error>> {
    if db_type == "postgresql" {
        if find_dump_util(db_type) {
            postgresql_backup(connection_string, backup_location, datetime)
        } else {
            Err(Box::try_from("pg_dump utility is not installed or not added to the environment variable").unwrap())
        }
    } else if db_type == "mysql" || db_type == "mariadb" {
        if find_dump_util(db_type) {
            mariadb_backup(connection_string, backup_location, datetime)
        } else {
            Err(Box::try_from("mysqldump utility is not installed or not added to the environment variable").unwrap())
        }
    } else if db_type == "sqlite" {
        if find_dump_util(db_type) {
            sqlite_backup(connection_string, backup_location, datetime)
        } else {
            Err(Box::try_from("sqlite3 utility is not installed or not added to the environment variable").unwrap())
        }

    } else {
        Err("Unsupported database type".into())
    }
}

fn postgresql_backup(connection_string: String, backup_location: &String, datetime: String) -> Result<(), Box<dyn Error>> {
    let backup_process:Output = Command::new("pg_dump")
        .arg("--dbname=".to_owned() + &connection_string)
        .arg("--format=custom")
        .arg(format!("--file={}{}-db.dump", backup_location, datetime))
        .output()?;

    if backup_process.status.success() {
        eprintln!("Database backup successfully completed. Dump stored at next path {}{}-db.dump", backup_location, datetime);
        Ok(())
    } else {
        let error = format!("Database backup not created. {}", String::from_utf8_lossy(&backup_process.stderr));
        Err(error.into())
    }
}

fn mariadb_backup(connection_string: String, backup_location: &String, datetime: String) -> Result<(), Box<dyn Error>> {
    let connection_string = format!("{} --result-file={}{}-db.sql", connection_string, backup_location, datetime);

    let backup_process = Command::new("mysqldump")
        .args(connection_string.split_whitespace().collect::<Vec<_>>())
        .output()?;

    if backup_process.status.success() {
        eprintln!("Database backup successfully completed. Dump stored at next path {}{}-db.sql", backup_location, datetime);
        Ok(())
    } else {
        let error = format!("Database backup not created. {}", String::from_utf8_lossy(&backup_process.stderr));
        Err(error.into())
    }
}

fn sqlite_backup(connection_string: String, backup_location: &String, datetime: String) -> Result<(), Box<dyn Error>> {
    let db_exist = File::open(format!("{}db.sqlite3", connection_string));
    if db_exist.is_ok() {
        let backup_process = Command::new("sqlite3")
            .arg(format!("{}db.sqlite3", connection_string))
            .arg(format!(".backup {}{}-db.sqlite3", backup_location, datetime))
            .output()?;

        if backup_process.status.success() {
            eprintln!("Database backup successfully completed. Dump stored at next path {}{}-db.sqlite3", backup_location, datetime);
            Ok(())
        } else {
            let error = format!("Database backup not created. {}", String::from_utf8_lossy(&backup_process.stderr));
            Err(error.into())
        }
    } else {
        Err("Database file not found".into())
    }
}

fn find_dump_util(db_type: String) -> bool {
    let mut status: bool = false;

    match db_type.as_ref() {
        "postgresql" => {
            let check = Command::new("which")
                .arg("pg_dump")
                .output()
                .expect("");

            if check.status.success() {
                status = true
            }
        }
        "mariadb" => {
            let check = Command::new("which")
                .arg("mysqldump")
                .output()
                .expect("");

            if check.status.success() {
                status = true
            }
        }
        "mysql" => {
            let check = Command::new("which")
                .arg("mysqldump")
                .output()
                .expect("");

            if check.status.success() {
                status = true
            }
        }
        "sqlite" => {
            let check = Command::new("which")
                .arg("sqlite3")
                .output()
                .expect("");

            if check.status.success() {
                status = true
            }
        }
        _ => {}
    };

    status
}

src/middleware.rs
use std::fs;
use std::fs::File;
use std::io::{Read, Write};
use walkdir::DirEntry;
use zip::unstable::write::FileOptionsExt;
use zip::write::FileOptions;

pub fn delete_temp_dir(temp_directory: &str) -> Result<(), std::io::Error> {
    fs::remove_dir_all(temp_directory)
}

pub fn zip_dir<T>(
    items: &mut dyn Iterator<Item = DirEntry>,
    directory: &str,
    archive_file: T,
    key: &str,
) -> zip::result::ZipResult<()>
    where
        T: Write + std::io::Seek,
{
    let mut zip = zip::ZipWriter::new(archive_file);
    let options = FileOptions::default()
        .compression_method(zip::CompressionMethod::Deflated)
        .unix_permissions(0o755)
        .with_deprecated_encryption(key.as_ref());

    let mut buffer = Vec::new();
    for entry in items {
        let path = entry.path();
        let name = path.strip_prefix(std::path::Path::new(directory)).unwrap();

        if path.is_file() {
            #[allow(deprecated)]
            zip.start_file_from_path(name, options)?;
            let mut f = File::open(path)?;

            f.read_to_end(&mut buffer)?;
            zip.write_all(&buffer)?;
            buffer.clear();
        } else if !name.as_os_str().is_empty() {
            #[allow(deprecated)]
            zip.add_directory_from_path(name, options)?;
        }
    }
    zip.finish()?;
    Ok(())
}

src/config.rs
use std::env;
use std::process::exit;
use std::fs;
use std::error::Error;
use std::path::Path;
use serde::Deserialize;

const CONFIG_DEFAULT_PATH: &str = "config.toml";

#[derive(Deserialize)]
pub struct Config {
    pub vaultwarden_data: String,
    pub backup_location: String,
    encrypt_data: Option<bool>,
    encrypt_key: Option<String>,
    exclude_files: Option<String>,
    db: Database,
}

#[derive(Deserialize)]
struct Database {
    db_type: String,
    username: Option<String>,
    password: Option<String>,
    db_name: Option<String>,
    host: Option<String>,
    port: Option<i64>,
}

impl Config {
    pub fn new() -> Config {
        let config = match read_config_file() {
            Ok(config) => config,
            Err(err) => {
                eprintln!("{}", err);
                exit(1);
            }
        };

        config
    }

    pub fn get_encrypt_status(&self) -> bool {
        self.encrypt_data.unwrap_or(false)
    }

    pub fn get_encrypt_key(&self) -> String {
        self.encrypt_key.clone().unwrap_or("".to_string())
    }

    fn get_db(&self) -> &Database {
        &self.db
    }

    pub fn get_db_type(&self) -> String {
        let db = self.get_db();
        db.db_type.clone()
    }

    pub fn db_connection_string(&self) -> String {
        let mut connection_string = String::new();
        let db = self.get_db();

        if db.db_type == "postgresql" {
            let mut credential = String::new();

            if let (Some(username), Some(password)) = (&db.username, &db.password) {
                credential.push_str(&format!("{}:{}@", username, password));
            }


            if !credential.is_empty() {
                connection_string.push_str(&format!(
                    "postgresql://{}{}:{}/{}",
                    credential,
                    db.host.as_ref().unwrap(),
                    db.port.unwrap_or(5432),
                    db.db_name.as_ref().unwrap())
                );
            } else {
                connection_string.push_str(&format!(
                    "postgresql://{}:{}/{}",
                    db.host.as_ref().unwrap(),
                    db.port.unwrap_or(5432),
                    db.db_name.as_ref().unwrap())
                );
            }

        } else if db.db_type == "mysql" || db.db_type == "mariadb" {
            if !db.username.is_none() && !db.password.is_none() {
                connection_string.push_str(&format!(
                    "--user={} --password={} --host={} --port={} {}",
                    db.username.as_ref().unwrap(),
                    db.password.as_ref().unwrap(),
                    db.host.as_ref().unwrap(),
                    db.port.unwrap_or(3306),
                    db.db_name.as_ref().unwrap()));
            } else {
                connection_string.push_str(&format!(
                    "--host={} --port={} {}",
                    db.host.as_ref().unwrap(),
                    db.port.unwrap_or(3306),
                    db.db_name.as_ref().unwrap()));
            }

        } else if db.db_type == "sqlite" {
            connection_string.push_str(&self.vaultwarden_data)
        }

        connection_string
    }

    pub fn get_backup_location(&self) -> Result<&String, Box<dyn Error>> {
        let backup_location = Path::new(&self.backup_location);
        if !backup_location.exists() || !backup_location.is_dir() {
            return Err("Backup folder does not exist".into());
        }

        Ok(&self.backup_location)
    }

    pub fn get_exclude_files(&self) -> Vec<String> {
        if let Some(excluded_files) = &self.exclude_files {
            excluded_files.split(',').map(|f| f.trim().to_string()).collect()
        } else {
            Vec::new()
        }
    }
}

pub fn get_config_path() -> String {
    match env::var("CONFIG_PATH") {
        Ok(config_path) => config_path,
        Err(_) => String::from(CONFIG_DEFAULT_PATH)
    }
}

pub fn read_config_file() -> Result<Config, String> {
    let config_path = get_config_path();

    let toml_content = fs::read_to_string(config_path)
        .map_err(|err| format!("Failed to open config file: {}", err))?;

    let config: Config = toml::from_str(&toml_content)
        .map_err(|err| format!("Failed to parse config file: {}", err.message()))?;

    Ok(config)
}